File: 241017/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.cpp
#include <iostream>
#include <vector>
using namespace std;
#define MAX_NM 100

int n, m;
int grid[MAX_NM][MAX_NM];
bool visited[MAX_NM][MAX_NM];

bool InRange(int x, int y) {
    return 0<=x && x<n && 0<=y && y<m;
}

bool CanGo(int x, int y) {
    if(!InRange(x, y)) return false;
    if(visited[x][y] || grid[x][y] == 0) return false;

    return true;
}

void DFS(int x, int y) {
    int dx[2] = {1, 0};
    int dy[2] = {0, 1};

    for(int d=0; d<2; d++) {
        int nx = x+dx[d];
        int ny = y+dy[d];

        if(CanGo(nx, ny)) {
            visited[nx][ny] = 1;
            DFS(nx, ny);
        }
    }
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n>>m;
    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            cin >> grid[i][j];
        }
    }

    visited[0][0] = 1;
    DFS(0, 0);

    cout << visited[n-1][m-1];
    return 0;
}

File: 241031/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.py
n, m = tuple(map(int, input().split()))
grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

visited = [
    [0 for _ in range(m)]
    for _ in range(n)
]

def in_range(x, y):
    return 0<=x and x<n and 0<=y and y<m

def can_go(x, y):
    if not in_range(x, y):
        return False
    if grid[x][y] == 0 or visited[x][y]:
        return False
    
    return True

def dfs(x, y):
    dxs, dys = [0, 1], [1, 0]

    for dx, dy in zip(dxs, dys):
        nx, ny = x+dx, y+dy

        if can_go(nx, ny):
            visited[nx][ny] = 1
            dfs(nx, ny)


visited[0][0] = True
dfs(0, 0)

print(visited[n-1][m-1])